File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char>st;
        for(int i=0;i<s.size();i++){
            char ch=s[i];

            if(ch=='(' || ch=='{' || ch=='[') 
                st.push(ch);
            else{
                if(st.empty()) 
                    return false;
                else{
                    char matching=st.top();
                    if((ch==')' && matching=='(') || (ch=='}' && matching=='{' )|| (ch==']' && matching=='[')) {
                        st.pop();
                    }
                    else return false;
                }
            }
        }
        if(st.size()==0) return true;
        return false;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for x in s:
            if x == "}" or x == ")" or x == "]":
                if len(stack) == 0:
                    return False
                if x == "}":
                    if stack[-1] == "{":
                        stack.pop()
                    else:
                        return False
                if x == "]":
                    if stack[-1] == "[":
                        stack.pop()
                    else:
                        return False
                if x == ")":
                    if stack[-1] == "(":
                        stack.pop()
                    else:
                        return False
            else:
                stack.append(x)

        if len(stack) > 0:
            return False

        return True
